### ISSUE TITLE
Missed optimization for .rem_euclid(2^n) on signed integer types.

### ISSUE THREAD

**Issue by HypheX** [rust-lang/rust#115855]
When calling `.rem_euclid(n)` on a signed integer type, it fails to apply the optimization for powers of 2, where it should do a bitwise AND operation instead.

## Missed optimization
```rs
pub fn rem(n: i32) -> i32 {
    n.rem_euclid(2)
}
```
```
example::rem:
        mov     eax, edi
        shr     eax, 31
        add     eax, edi
        and     eax, -2
        sub     edi, eax
        mov     eax, 1
        cmovns  eax, edi
        ret
```

## How it should be optimized
```rs
pub fn rem(n: i32) -> i32 {
    n & 1
}
```
```
example::rem:
        mov     eax, edi
        and     eax, 1
        ret
```

<details>
<summary> Testing the optimization: </summary>

## n = 2
?play
```rs
let mut passed = true;
for i in (std::i16::MIN..=std::i16::MAX) {
    if (i&1) != (i.rem_euclid(2)) {
        passed = false;
        break
    }
}
println!("{passed}");
```
```rs
true
```
## n = 1
?play
```rs
let mut passed = true;
for i in (std::i16::MIN..=std::i16::MAX) {
    if (i&0) != (i.rem_euclid(1)) {
        passed = false;
        break
    }
}
println!("{passed}");
```
```rs
true
```
## n = 4
?play
```rs
let mut passed = true;
for i in (std::i16::MIN..=std::i16::MAX) {
    if (i&3) != (i.rem_euclid(4)) {
        passed = false;
        break
    }
}
println!("{passed}");
```
```rs
true
```
</details>

---

Curiously, the optimization is applied to unsigned integers just fine
```rs
pub fn rem(n: u32) -> u32 {
    n.rem_euclid(2) 
}
```
```
example::rem:
        mov     eax, edi
        and     eax, 1
        ret
```

**Comment by nikic** [comment]
If you are going to file an upstream issue yourself, please do link back to it so other people don't do duplicate work: https://github.com/llvm/llvm-project/issues/66417

Anyway, this issue should already be fixed in LLVM 18.

**Comment by fmease** [comment]
As an aside, the title should be *2^n* (powers of two) instead of *n^2* (squares).

**Comment by HypheX** [comment]
> As an aside, the title should be _2^n_ (powers of two) instead of _n^2_ (squares).
Thanks, wow, had a brainfart there.

**Comment by nikic** [comment]
Confirmed fixed by https://github.com/rust-lang/rust/pull/120055, needs codegen test. Godbolt: https://rust.godbolt.org/z/cf3Ej1Mq3

**Comment by HypheX** [comment]
Fixed in 1.78.0 (Updated to LLVM 18)
